pkg/runtime/dynamic_components_loader.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2022 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package runtime
+
+import (
+	"path/filepath"
+
+	"github.com/dapr/dapr/pkg/components"
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/errors"
+
+	componentsV1alpha1 "github.com/dapr/dapr/pkg/apis/components/v1alpha1"
+)
+
+func (a *DaprRuntime) watchPathForDynamicLoading() {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.Errorf("unable to create watcher for dynamic components, dynamic loading of components will not be supported: %w", err)
+		return
+	}
+	defer watcher.Close()
+	err = watcher.Add(a.runtimeConfig.Standalone.ComponentsPath)
+	if err != nil {
+		log.Errorf("unable to watch components directory: %s , err: %w , dynamic loading of components will not be supported", a.runtimeConfig.Standalone.ComponentsPath, err)
+		return
+	}
+	for {
+		select {
+		case event, ok := <-watcher.Events:
+			if !ok {
+				return
+			}
+			log.Debug("observed a file event in components directory:", event)
+			if event.Op == fsnotify.Create || event.Op == fsnotify.Write {
+				err := a.loadDynamicComponents(event.Name)
+				if err != nil {
+					log.Errorf("failed to load components from file: %s err: %w", event.Name, err)
+				}
+			} else if event.Op == fsnotify.Remove || event.Op == fsnotify.Rename {
+				for _, comp := range a.dynamicComponents[DynamicComponentsManifest(event.Name)] {
+					log.Warnf("file: %s deleted, component: %s will not be loaded on sidecar restart.", filepath.Base(event.Name), comp.Name)
```

Comment:
Nit: `will not be loaded when daprd is restarted` (because technically "sidecar" is a K8s-only concept)